add new model blog (#1041)
huggingface/blog: if.md

---
title: "Running IF with 🧨 diffusers on a Free Tier Google Colab"
thumbnail: /blog/assets/if/thumbnail.jpg
authors:

> - user: DeepFloyd 

Reviewer: Added users instead

  guest: true
- user: williamberman
- user: patrickvonplaten
---

# Running IF with 🧨 diffusers on a Free Tier Google Colab {#running-if-with--diffusers-on-a-free-tier-google-colab}

<!-- {blog_metadata} -->
<!-- {authors} -->

> ***TL;DR**: We show how to run one of the most powerful open-source text

Reviewer 2: Everything is in bold. I think the three asterisks is messing this up

Author: https://github.com/huggingface/blog/pull/1041/commits/53ab3d057d76ad1b2b643a0b2999b41848783407

well as GPU VRAM constrained (15 GB RAM for T4) which makes running the
whole \>10B IF model challenging!

Let\'s map out the size of IF\'s model components in full float32
precision:

- [T5-XXL Text Encoder](https://huggingface.co/DeepFloyd/IF-I-IF-v1.0/tree/main/text_encoder): 20GB
- [Stage 1 UNet](https://huggingface.co/DeepFloyd/IF-I-IF-v1.0/tree/main/unet): 17.2 GB
- [Stage 2 Super Resolution UNet](https://huggingface.co/DeepFloyd/IF-II-L-v1.0/blob/main/pytorch_model.bin): 2.5 GB
- [Stage 3 Super Resolution Model](https://huggingface.co/stabilityai/stable-diffusion-x4-upscaler): 3.4 GB

There is no way we can run the model in float32 as the T5 and Stage 1
UNet weights are each larger than the available CPU RAM.

In float16, the component sizes are 11GB, 8.6GB and 1.25GB for T5,
Stage1 and Stage2 UNets respectively which is doable for the GPU, but
we\'re still running into CPU memory overflow errors when loading the T5
(some CPU is occupied by other processes).

Therefore, we lower the precision of T5 even more by using
`bitsandbytes` 8bit quantization which allows to save the T5 checkpoint
with as little as [8

> GB](https://huggingface.co/DeepFloyd/IF-I-IF-v1.0/blob/main/text_encoder/model.8bit.safetensors).

Reviewer: Added a section to the top

Now that each components fits individually into both CPU and GPU memory,
we need to make sure that components have all the CPU and GPU memory for
themselves when needed.

Diffusers supports modularly loading individual components i.e. we can
load the text encoder without loading the unet. This modular loading
will ensure that we only load the component we need at a given step in
the pipeline to avoid exhausting the available CPU RAM and GPU VRAM.

Let\'s give it a try 🚀

![t2i_64](assets/if/t2i_64.png)

## Available resources

The free-tier Google Colab comes with around 13 GB CPU RAM:

``` python
!grep MemTotal /proc/meminfo
```

```bash
MemTotal:       13297192 kB
```

And an NVIDIA T4 with 15 GB VRAM:

``` python
!nvidia-smi
```

```bash

> Sun Apr 23 23:14:19 2023       

Author: discussed in slack, will leave as is

you are having issues, please double check and upgrade versions.

``` python
! pip install --upgrade \
  diffusers~=0.16 \
  transformers~=4.28 \
  safetensors~=0.3 \
  sentencepiece~=0.1 \
  accelerate~=0.18 \
  bitsandbytes~=0.38 \
  torch~=2.0 -q
```

## 1. Text-to-image generation

We will walk step by step through text-to-image generation with IF using
Diffusers. We will explain briefly APIs and optimizations, but more
in-depth explanations can be found in the official documentation for
[Diffusers](https://huggingface.co/docs/diffusers/index),
[Transformers](https://huggingface.co/docs/transformers/index),
[Accelerate](https://huggingface.co/docs/accelerate/index), and
[bitsandbytes](https://github.com/TimDettmers/bitsandbytes).

### 1.1 Load text encoder

We will load T5 using 8bit quantization. Transformers directly supports
[bitsandbytes](https://huggingface.co/docs/transformers/main/en/main_classes/quantization#load-a-large-model-in-8bit)
through the `load_in_8bit` flag.

The flag `variant="8bit"` will download pre-quantized weights.

We also use the `device_map` flag to allow `transformers` to offload
model layers to the CPU or disk. Transformers big modeling supports

> arbitrary device maps which can be used to separately load model

Reviewer 2: ```suggestion
arbitrary device maps, which can be used to separately load model
```

parameters directly to available devices. Passing `"auto"` will
automatically create a device map. See the `transformers`
[docs](https://huggingface.co/docs/accelerate/usage_guides/big_modeling#designing-a-device-map)
for more information.

``` python
from transformers import T5EncoderModel

text_encoder = T5EncoderModel.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0",
    subfolder="text_encoder", 
    device_map="auto", 
    load_in_8bit=True, 
    variant="8bit"
)
```

### 1.2 Create text embeddings

The Diffusers API for accessing diffusion models is the
`DiffusionPipeline` class and its subclasses. Each instance of
`DiffusionPipeline` is a fully self contained set of methods and models
for running diffusion networks. We can override the models it uses by
passing alternative instances as keyword arguments to `from_pretrained`.

> In this case, we pass `None` for the `unet` argument so no UNet will be

Reviewer 2: ```suggestion
In this case, we pass `None` for the `unet` argument, so no UNet will be
```

loaded. This allows us to run the text embedding portion of the
diffusion process without loading the UNet into memory.

``` python
from diffusers import DiffusionPipeline

pipe = DiffusionPipeline.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0", 
    text_encoder=text_encoder, # pass the previously instantiated 8bit text encoder
    unet=None, 
    device_map="auto"
)
```

IF also comes with a super resolution pipeline. We will save the prompt

> embeddings so that we can later directly pass them to the super

Reviewer 2: ```suggestion
embeddings so we can later directly pass them to the super
```

resolution pipeline. This will allow the super resolution pipeline to be
loaded **without** a text encoder.

Instead of [an astronaut just riding a

> horse](https://huggingface.co/blog/stable_diffusion), let\'s hand him a

Reviewer 2: ```suggestion
horse](https://huggingface.co/blog/stable_diffusion), let\'s hand them a
```

sign as well!

Let\'s define a fitting prompt:

``` python

> prompt = 'a photograph of an astronaut riding a horse holding a sign that says "Pixel\'s in space"'

Reviewer 2: ```suggestion
prompt = "a photograph of an astronaut riding a horse holding a sign that says Pixel's in space"
```

```

and run it through the 8bit quantized T5 model:

``` python
prompt_embeds, negative_embeds = pipe.encode_prompt(prompt)
```

### 1.3 Free memory

Once the prompt embeddings have been created. We do not need the text
encoder anymore. However, it is still in memory on the GPU. We need to
remove it so that we can load the UNet.

> It\'s non-trivial to free PyTorch memory. We must garbage collect the

Reviewer 2: ```suggestion
It's non-trivial to free PyTorch memory. We must garbage-collect the
```

Python objects which point to the actual memory allocated on the GPU.

> First, use the python keyword `del` to delete all python objects

Reviewer 2: Here and in other instances
```suggestion
First, use the Python keyword `del` to delete all Python objects
```

referencing allocated GPU memory

``` python
del text_encoder
del pipe
```

> The deletion of the python object is not enough to free the GPU memory.

Reviewer 2: ```suggestion
Deleting the python object is not enough to free the GPU memory.
```

Garbage collection is when the actual GPU memory is freed.

Additionally, we will call `torch.cuda.empty_cache()`. This method
isn\'t strictly necessary as the cached cuda memory will be immediately
available for further allocations. Emptying the cache allows us to

> verify in the colab UI that the memory is available.

Reviewer 2: ```suggestion
verify in the Colab UI that the memory is available.
```

We\'ll use a helper function `flush()` to flush memory.

``` python
import gc
import torch

def flush():
    gc.collect()
    torch.cuda.empty_cache()
```

and run it

``` python
flush()
```

### 1.4 Stage 1: The main diffusion process

With our now available GPU memory, we can re-load the
`DiffusionPipeline` with only the UNet to run the main diffusion
process.

The `variant` and `torch_dtype` flags are used by Diffusers to download
and load the weights in 16 bit floating point format.

``` python
pipe = DiffusionPipeline.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0", 
    text_encoder=None, 
    variant="fp16", 
    torch_dtype=torch.float16, 
    device_map="auto"
)
```

Often, we directly pass the text prompt to `DiffusionPipeline.__call__`.
However, we previously computed our text embeddings which we can pass
instead.

IF also comes with a super resolution diffusion process. Setting
`output_type="pt"` will return raw PyTorch tensors instead of a PIL

> image. This way we can keep the Pytorch tensors on GPU and pass them

Reviewer 2: ```suggestion
image. This way, we can keep the PyTorch tensors on GPU and pass them
```

the previous PyTorch tensor output and pass it to the tage 3 pipeline as
`image=image`.

💡 **Note**: The x4 Upscaler does not use T5 and has [its own text
encoder](https://huggingface.co/stabilityai/stable-diffusion-x4-upscaler/tree/main/text_encoder).
Therefore, we cannot use the previously created prompt embeddings and
instead must pass the original prompt.

``` python
pil_image = pipe(prompt, generator=generator, image=image).images
```

Unlike the IF pipelines, the IF watermark will not be added by default
to outputs from the Stable Diffusion x4 upscaler pipeline.

We can instead manually apply the watermark.

``` python
from diffusers.pipelines.deepfloyd_if import IFWatermarker

watermarker = IFWatermarker.from_pretrained("DeepFloyd/IF-I-IF-v1.0", subfolder="watermarker")
watermarker.apply_watermark(pil_image, pipe.unet.config.sample_size)
```

View output image

``` python
pil_image[0]
```

![t2i_upscaled_2](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/if/t2i_upscaled_2.png)

Et voila! A beautiful 1024x1024 image in a free-tier Google Colab.

We have shown how 🧨 diffusers makes it easy to decompose and modularly

> load resource intensive diffusion models.

Reviewer 2: ```suggestion
load resource-intensive diffusion models.
```

💡 **Note**: We don\'t recommend using the above setup in production.
8bit quantization, manual de-allocation of model weights, and disk
offloading all trade off memory for time (i.e., inference speed). This
can be especially noticable if the diffusion pipeline is re-used. In
production, we recommend using a 40GB A100 with all model components
left on the GPU. See [**the official IF
demo**](https://huggingface.co/spaces/DeepFloyd/IF).

## 2. Image variation

The same IF checkpoints can also be used for text guided image variation

> and inpainting. The core diffusion process is the same as text to image

Reviewer 2: ```suggestion
and inpainting. The core diffusion process is the same as text-to-image
```

generation except the initial noised image is created from the image to
be varied or inpainted.

To run image variation, load the same checkpoints with
`IFImg2ImgPipeline.from_pretrained()` and
`IFImg2ImgSuperResolution.from_pretrained()`.

The APIs for memory optimization are all the same!

Let\'s free the memory from the previous section.

``` python
del pipe
flush()
```

For image variation, we start with an initial image that we want to
adapt.

For this section, we will adapt the famous \"Slaps Roof of Car\" meme.
Let\'s download it from the internet.

``` python
import requests

url = "https://i.kym-cdn.com/entries/icons/original/000/026/561/car.jpg"
response = requests.get(url)
```

and load it into a PIL Image

``` python
from PIL import Image
from io import BytesIO

original_image = Image.open(BytesIO(response.content)).convert("RGB")
original_image = original_image.resize((768, 512))
original_image
```

![iv_sample](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/if/iv_sample.png)

> The image variation pipeline take both PIL images and raw tensors. View

Reviewer 2: ```suggestion
The image variation pipeline takes both PIL images and raw tensors. View
```

> the docstrings for more indepth documentation on expected inputs (TODO

Reviewer 2: TODO

link to docstring)

### 2.1 Text Encoder

Image variation is guided by text, so we can define a prompt and encode
it with T5\'s Text Encoder.

Again we load the text encoder into 8bit precision.

``` python
from transformers import T5EncoderModel

text_encoder = T5EncoderModel.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0",
    subfolder="text_encoder", 
    device_map="auto", 
    load_in_8bit=True, 
    variant="8bit"
)
```

For image variation, we load the checkpoint with

> [`IFImg2ImgPipeline(TODO <link>)`](). When using

Reviewer 2: TODO here and below

`DiffusionPipeline.from_pretrained(...)`, checkpoints are loaded into
their default pipeline. The default pipeline for the IF is the
text-to-image [`IFPipeline(TODO <link>`](). When loading checkpoints
with a non-default pipeline, the pipeline must be explicitly specified.

``` python
from diffusers import IFImg2ImgPipeline

pipe = IFImg2ImgPipeline.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0", 
    text_encoder=text_encoder, 
    unet=None, 
    device_map="auto"
)
```

Let\'s turn our salesman into an anime character.

``` python
prompt = "anime style"
```

As before, we create the text embeddings with T5

``` python
prompt_embeds, negative_embeds = pipe.encode_prompt(prompt)
```

and free GPU and CPU memory.

> First remove the Python pointers

Reviewer 2: ```suggestion
First, remove the Python pointers
```

flush()
```

### 2.2 Stage 1: The main diffusion process 

Next, we only load the stage 1 UNet weights into the pipeline object,
just like we did in the previous section.

``` python
pipe = IFImg2ImgPipeline.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0", 
    text_encoder=None, 
    variant="fp16", 
    torch_dtype=torch.float16, 
    device_map="auto"
)
```

The image variation pipeline requires both the original image and the
prompt embeddings.

We can optionally use the `strength` argument to configure the amount of
variation. `strength` directly controls the amount of noise added.
Higher strength means more noise which means more variation.

``` python
generator = torch.Generator().manual_seed(0)
image = pipe(
    image=original_image,
    prompt_embeds=prompt_embeds,
    negative_prompt_embeds=negative_embeds, 
    output_type="pt",
    generator=generator,
).images
```

Let\'s check the intermediate 64x64 again.

``` python
pil_image = pt_to_pil(image)
pipe.watermarker.apply_watermark(pil_image, pipe.unet.config.sample_size)

pil_image[0]
```

![iv_sample_1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/if/iv_sample_1.png)

> Looks good, we can free the memory and upscale the image again.

Reviewer 2: ```suggestion
Looks good! We can free the memory and upscale the image again.
```

``` python
del pipe
flush()
```

### 2.3 Stage 2: Super Resolution

For super resolution, load the checkpoint with
`IFImg2ImgSuperResolutionPipeline` and the same checkpoint as before.

``` python
from diffusers import IFImg2ImgSuperResolutionPipeline

pipe = IFImg2ImgSuperResolutionPipeline.from_pretrained(
    "DeepFloyd/IF-II-L-v1.0", 
    text_encoder=None, 
    variant="fp16", 
    torch_dtype=torch.float16, 
    device_map="auto"
)
```
💡 **Note**: The image variation super resolution pipeline requires the
generated image as well as the original image.

You can also use the Stable Diffusion x4 upscaler on this image. Feel
free to try it out using the code snippets in section 1.6.

``` python
image = pipe(
    image=image,
    original_image=original_image,
    prompt_embeds=prompt_embeds,
    negative_prompt_embeds=negative_embeds, 
    generator=generator,
).images[0]
image
```

![iv_sample_2](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/if/iv_sample_2.png)

Nice! Let\'s free the memory and look at the final inpainting pipelines.

``` python
del pipe
flush()
```

## 3. Inpainting

> The IF inpainting pipeline is the same as the image variation except
> only a select area of the image is denoised.

Reviewer 2: ```suggestion
The IF inpainting pipeline is the same as the image variation, except
only a select area of the image is denoised.
```

> We specify the area to inpatint with an image mask.

Reviewer 2: ```suggestion
We specify the area to inpaint with an image mask.
```

# Make sure to create the image in mode 'L'
# meaning single channel grayscale
example_mask = Image.fromarray(example_mask, mode='L')

example_mask
```

![masking_by_hand](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/if/masking_by_hand.png)

Now we can start inpainting 🎨🖌

### 3.1. Text Encoder 

Again, we load the text encoder first

``` python
from transformers import T5EncoderModel

text_encoder = T5EncoderModel.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0",
    subfolder="text_encoder", 
    device_map="auto", 
    load_in_8bit=True, 
    variant="8bit"
)
```

This time, we initialize the `IFInpaintingPipeline` in-painting pipeline
with the text encoder weights.

``` python
from diffusers import IFInpaintingPipeline

pipe = IFInpaintingPipeline.from_pretrained(
    "DeepFloyd/IF-I-IF-v1.0", 
    text_encoder=text_encoder, 
    unet=None, 
    device_map="auto"
)
```

Alright, let\'s have the man advertise for more layers instead.

``` python
prompt = 'the text, "just stack more layers"'
```

Having defined the prompt, we can create the prompt embeddings

``` python
prompt_embeds, negative_embeds = pipe.encode_prompt(prompt)
```

> Just like before we free the memory

Reviewer 2: ```suggestion
Just like before, we free the memory
```

``` python
del text_encoder
del pipe
flush()
```

### 3.2 Stage 1: The main diffusion process 

> Just like before we now load the stage 1 pipeline with only the UNet.

Reviewer 2: ```suggestion
Just like before, we now load the stage 1 pipeline with only the UNet.
```

![inpainted_output](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/if/inpainted_output.png)

Looks good! The text is pretty consistent!

Let\'s free the memory so we can upscale the image

``` python
del pipe
flush()
```

### 3.3 Stage 2: Super Resolution 

For super resolution, load the checkpoint with
`IFInpaintingSuperResolutionPipeline`.

``` python
from diffusers import IFInpaintingSuperResolutionPipeline

pipe = IFInpaintingSuperResolutionPipeline.from_pretrained(
    "DeepFloyd/IF-II-L-v1.0", 
    text_encoder=None, 
    variant="fp16", 
    torch_dtype=torch.float16, 
    device_map="auto"
)
```

The inpainting super resolution pipeline requires the generated image,
the original image, the mask image, and the prompt embeddings.

Let\'s do a final denoising run.

``` python
image = pipe(
    image=image,
    original_image=original_image,
    mask_image=mask_image,
    prompt_embeds=prompt_embeds,
    negative_prompt_embeds=negative_embeds, 
    generator=generator,
).images[0]
image
```

![inpainted_final_output](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/if/inpainted_final_output.png)

> Nice, the model managed to generate text without making a single

Reviewer 2: ```suggestion
Nice, the model generated text without making a single
```

spelling error!

## Conclusion

> IF in 32 bit floating point precision uses 40 GB of weights in total. We

Reviewer 2: ```suggestion
IF in 32-bit floating point precision uses 40 GB of weights in total. We
```

showed how using only open source models and libraries, IF can be ran on
a free-tier Google Colab instance.

The ML ecosystem benefits deeply from the sharing of open tools and open

> models. This notebook alone used models from DeepFloyd, StabilityAI, and

Reviewer 2: Also Google for T5

> models. This notebook alone used models from DeepFloyd, StabilityAI, and

Reviewer 2: What model from LAION was used?
